File: jetbrains-plugin/src/main/kotlin/com/claudecodeplus/server/services/MessageContentParser.kt
package com.claudecodeplus.server.services

import com.claudecodeplus.sdk.types.*
import com.intellij.openapi.diagnostic.Logger
import java.io.File
import java.util.Base64

/**
 * 消息内容解析器
 * 
 * 功能：
 * 1. 解析用户消息中的 @ 引用
 * 2. 识别图片路径并转换为 ImageBlock
 * 3. 保留普通文件引用给 Claude Code 处理
 */
object MessageContentParser {
    private val logger = Logger.getInstance(MessageContentParser::class.java)
    
    // 图片文件扩展名
    private val IMAGE_EXTENSIONS = setOf("png", "jpg", "jpeg", "gif", "bmp", "webp", "svg")
    
    /**
     * 解析消息文本，提取 @ 引用并转换为 ContentBlock 列表
     * 
     * @param text 用户输入的消息文本
     * @return ContentBlock 列表（TextBlock 和 ImageBlock）
     */
    fun parseMessageContent(text: String): List<ContentBlock> {
        val blocks = mutableListOf<ContentBlock>()
        val pattern = """@([^\s]+)""".toRegex() // 匹配 @ 后面的路径（到空格为止）
        
        var lastIndex = 0
        
        pattern.findAll(text).forEach { match ->
            // 添加前面的文本块
            if (match.range.first > lastIndex) {
                val textContent = text.substring(lastIndex, match.range.first)
                if (textContent.isNotBlank()) {
                    blocks.add(TextBlock(textContent))
                }
            }
            
            val path = match.groupValues[1]
            
            // 判断是否为图片路径
            if (isImagePath(path)) {
                // ImageBlock 在当前 SDK 版本中不存在，暂时作为文本处理
                blocks.add(TextBlock("[图片: ${File(path).name}]"))
            } else {
                // 普通文件引用，保留原样（Claude Code 会处理）
                blocks.add(TextBlock("@$path"))
            }
            
            lastIndex = match.range.last + 1
        }
        
        // 添加剩余文本
        if (lastIndex < text.length) {
            val textContent = text.substring(lastIndex)
            if (textContent.isNotBlank()) {
                blocks.add(TextBlock(textContent))
            }
        }
        
        // 如果没有任何块，返回一个空文本块
        if (blocks.isEmpty()) {
            blocks.add(TextBlock(text))
        }
        
        return blocks
    }
    
    /**
     * 判断路径是否为图片文件
     */
    private fun isImagePath(path: String): Boolean {
        val extension = path.substringAfterLast('.', "").lowercase()
        return IMAGE_EXTENSIONS.contains(extension)
    }
    
    /**
     * 将图片路径转换为 ImageBlock
     *
     * ImageBlock 在当前 SDK 版本中不存在，此方法已废弃
     *
     * @param path 图片文件路径（绝对路径或相对路径）
     * @return ImageBlock 或 null（如果文件不存在或读取失败）
     */
    @Deprecated("ImageBlock type does not exist in current SDK version")
    private fun convertToImageBlock(path: String): Any? {
        return try {
            val file = File(path)

            if (!file.exists() || !file.isFile) {
                logger.warn("Image file not found: $path")
                return null
            }

            // ImageBlock 在当前 SDK 版本中不存在，返回 null
            logger.warn("ImageBlock is not supported in current SDK version")
            return null

            // 以下代码已注释，因为 ImageBlock 和 ImageSource 类型不存在
            /*
            // 读取文件内容
            val fileBytes = file.readBytes()

            // 转换为 base64
            val base64Data = Base64.getEncoder().encodeToString(fileBytes)

            // 检测 MIME 类型
            val mimeType = detectMimeType(file)

            logger.info("Converted image to ImageBlock: ${file.name} (${fileBytes.size} bytes)")

            ImageBlock(
                source = ImageSource(
                    type = "base64",
                    mediaType = mimeType,
                    data = base64Data
                )
            )
            */
        } catch (e: Exception) {
            logger.error("Failed to convert image to ImageBlock: $path", e)
            null
        }
    }
    
    /**
     * 检测图片的 MIME 类型
     */
    private fun detectMimeType(file: File): String {
        return when (file.extension.lowercase()) {
            "png" -> "image/png"
            "jpg", "jpeg" -> "image/jpeg"
            "gif" -> "image/gif"
            "bmp" -> "image/bmp"
            "webp" -> "image/webp"
            "svg" -> "image/svg+xml"
            else -> "image/png" // 默认
        }
    }
}

File: vscode-extension/src/ide/util/messageContentParser.ts
/**
 * 消息内容解析器
 *
 * 功能：
 * 1. 解析用户消息中的 @ 引用
 * 2. 识别图片路径并转换为 ImageBlock
 * 3. 保留普通文件引用给 Claude Code 处理
 *
 * 对应 Kotlin 版本: jetbrains-plugin/src/main/kotlin/com/asakii/server/services/MessageContentParser.kt
 */

import * as fs from 'fs';
import * as path from 'path';
import { getLogger } from '../../logging';

const logger = getLogger('MessageContentParser');

// ==================== 类型定义 ====================

/**
 * 内容块基础类型
 */
export type ContentBlock = TextBlock | ImageBlock;

/**
 * 文本块
 */
export interface TextBlock {
    type: 'text';
    text: string;
}

/**
 * 图片源
 */
export interface ImageSource {
    type: 'base64' | 'url';
    mediaType: string;
    data: string;
}

/**
 * 图片块
 */
export interface ImageBlock {
    type: 'image';
    source: ImageSource;
}

// ==================== 常量 ====================

/** 图片文件扩展名 */
const IMAGE_EXTENSIONS = new Set([
    'png', 'jpg', 'jpeg', 'gif', 'bmp', 'webp', 'svg'
]);

/** MIME 类型映射 */
const MIME_TYPES: Record<string, string> = {
    'png': 'image/png',
    'jpg': 'image/jpeg',
    'jpeg': 'image/jpeg',
    'gif': 'image/gif',
    'bmp': 'image/bmp',
    'webp': 'image/webp',
    'svg': 'image/svg+xml'
};

// ==================== MessageContentParser 类 ====================

/**
 * 消息内容解析器
 */
export class MessageContentParser {
    /** @ 引用的正则表达式 */
    private static readonly AT_REFERENCE_PATTERN = /@([^\s]+)/g;

    /**
     * 解析消息文本，提取 @ 引用并转换为 ContentBlock 列表
     *
     * @param text 用户输入的消息文本
     * @param convertImages 是否将图片转换为 ImageBlock（默认 false，作为文本处理）
     * @returns ContentBlock 列表（TextBlock 和 ImageBlock）
     */
    static parseMessageContent(text: string, convertImages: boolean = false): ContentBlock[] {
        const blocks: ContentBlock[] = [];
        const pattern = new RegExp(this.AT_REFERENCE_PATTERN);
        
        let lastIndex = 0;
        let match: RegExpExecArray | null;

        while ((match = pattern.exec(text)) !== null) {
            // 添加前面的文本块
            if (match.index > lastIndex) {
                const textContent = text.substring(lastIndex, match.index);
                if (textContent.trim()) {
                    blocks.push(this.createTextBlock(textContent));
                }
            }

            const referencePath = match[1];

            // 判断是否为图片路径
            if (this.isImagePath(referencePath)) {
                if (convertImages) {
                    // 尝试转换为 ImageBlock
                    const imageBlock = this.convertToImageBlock(referencePath);
                    if (imageBlock) {
                        blocks.push(imageBlock);
                    } else {
                        // 转换失败，作为文本处理
                        const fileName = path.basename(referencePath);
                        blocks.push(this.createTextBlock(`[图片: ${fileName}]`));
                    }
                } else {
                    // 不转换，作为文本处理
                    const fileName = path.basename(referencePath);
                    blocks.push(this.createTextBlock(`[图片: ${fileName}]`));
                }
            } else {
                // 普通文件引用，保留原样（Claude Code 会处理）
                blocks.push(this.createTextBlock(`@${referencePath}`));
            }

            lastIndex = match.index + match[0].length;
        }

        // 添加剩余文本
        if (lastIndex < text.length) {
            const textContent = text.substring(lastIndex);
            if (textContent.trim()) {
                blocks.push(this.createTextBlock(textContent));
            }
        }

        // 如果没有任何块，返回一个包含完整文本的块
        if (blocks.length === 0) {
            blocks.push(this.createTextBlock(text));
        }

        return blocks;
    }

    /**
     * 判断路径是否为图片文件
     */
    static isImagePath(filePath: string): boolean {
        const extension = this.getExtension(filePath).toLowerCase();
        return IMAGE_EXTENSIONS.has(extension);
    }

    /**
     * 将图片路径转换为 ImageBlock
     *
     * @param imagePath 图片文件路径（绝对路径或相对路径）
     * @returns ImageBlock 或 null（如果文件不存在或读取失败）
     */
    static convertToImageBlock(imagePath: string): ImageBlock | null {
        try {
            if (!fs.existsSync(imagePath)) {
                logger.warn(`Image file not found: ${imagePath}`);
                return null;
            }

            const stats = fs.statSync(imagePath);
            if (!stats.isFile()) {
                logger.warn(`Path is not a file: ${imagePath}`);
                return null;
            }

            // 读取文件内容
            const fileBytes = fs.readFileSync(imagePath);

            // 转换为 base64
            const base64Data = fileBytes.toString('base64');

            // 检测 MIME 类型
            const mimeType = this.detectMimeType(imagePath);

            logger.info(`Converted image to ImageBlock: ${path.basename(imagePath)} (${fileBytes.length} bytes)`);

            return {
                type: 'image',
                source: {
                    type: 'base64',
                    mediaType: mimeType,
                    data: base64Data
                }
            };
        } catch (e) {
            logger.error(`Failed to convert image to ImageBlock: ${imagePath}`, e instanceof Error ? e : undefined);
            return null;
        }
    }

    /**
     * 检测图片的 MIME 类型
     */
    static detectMimeType(filePath: string): string {
        const extension = this.getExtension(filePath).toLowerCase();
        return MIME_TYPES[extension] ?? 'image/png'; // 默认 PNG
    }

    /**
     * 获取文件扩展名（不含点号）
     */
    private static getExtension(filePath: string): string {
        const lastDotIndex = filePath.lastIndexOf('.');
        if (lastDotIndex === -1 || lastDotIndex === filePath.length - 1) {
            return '';
        }
        return filePath.substring(lastDotIndex + 1);
    }

    /**
     * 创建文本块
     */
    private static createTextBlock(text: string): TextBlock {
        return {
            type: 'text',
            text: text
        };
    }

    /**
     * 提取消息中的所有 @ 引用路径
     *
     * @param text 消息文本
     * @returns 引用路径列表
     */
    static extractReferences(text: string): string[] {
        const references: string[] = [];
        const pattern = new RegExp(this.AT_REFERENCE_PATTERN);
        let match: RegExpExecArray | null;

        while ((match = pattern.exec(text)) !== null) {
            references.push(match[1]);
        }

        return references;
    }

    /**
     * 分离图片引用和文件引用
     *
     * @param text 消息文本
     * @returns 分离后的引用
     */
    static separateReferences(text: string): { images: string[]; files: string[] } {
        const references = this.extractReferences(text);
        const images: string[] = [];
        const files: string[] = [];

        for (const ref of references) {
            if (this.isImagePath(ref)) {
                images.push(ref);
            } else {
                files.push(ref);
            }
        }

        return { images, files };
    }

    /**
     * 将 ContentBlock 列表转换回纯文本
     *
     * @param blocks ContentBlock 列表
     * @returns 纯文本
     */
    static blocksToText(blocks: ContentBlock[]): string {
        return blocks
            .map(block => {
                if (block.type === 'text') {
                    return block.text;
                } else if (block.type === 'image') {
                    return '[Image]';
                }
                return '';
            })
            .join('');
    }
}

// ==================== 辅助函数 ====================

/**
 * 解析消息内容
 */
export function parseMessageContent(text: string, convertImages?: boolean): ContentBlock[] {
    return MessageContentParser.parseMessageContent(text, convertImages);
}

/**
 * 判断是否为图片路径
 */
export function isImagePath(filePath: string): boolean {
    return MessageContentParser.isImagePath(filePath);
}

/**
 * 提取消息中的所有引用
 */
export function extractReferences(text: string): string[] {
    return MessageContentParser.extractReferences(text);
}

/**
 * 检测 MIME 类型
 */
export function detectMimeType(filePath: string): string {
    return MessageContentParser.detectMimeType(filePath);
}
